abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+`response_code_details_size`, `response_body_data`, `response_body_size`,
+`additional_headers_map_data` and/or `additional_headers_size` point to invalid
+memory address.
+
+#### `proxy_resume_http_stream`
+
+* params:
+  - `i32 (proxy_stream_type_t) stream_type`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Resume processing of paused HTTP request or response (`stream_type`) in current
+context.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `stream_type`.
+
+Returns `InvalidOperation` when stream wasn't paused.
+
+#### `proxy_close_http_stream`
+
+* params:
+  - `i32 (proxy_stream_type_t) stream_type`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Close HTTP request or response (`stream_type`) in current context.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `stream_type`.
+
+Returns `InvalidOperation` when stream is already closed.
+
+
+## Buffers and maps
+
+### Functions exposed by the host environment
+
+#### `proxy_get_buffer`
+
+* params:
+  - `i32 (proxy_buffer_type_t) buffer_type`
+  - `i32 (size_t) offset`
+  - `i32 (size_t) max_size`
+  - `i32 (iovec_array_t*) return_buffer_data`
+  - `i32 (size_t*) return_buffer_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Returns content of the buffer `buffer_type`. At most `max_size` bytes will be
+returned, starting from the `offset`. Returned bytes are written into memory
+slices pointed by a list of vectors (`return_buffer_data`,
+`return_buffer_size`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `buffer_type`.
+
+Returns `Empty` when `buffer_type` doesn't contain any data.
+
+Returns `InvalidMemoryAccess` when `return_buffer_data` and/or
+`return_buffer_size` point to invalid memory address.
+
+#### `proxy_set_buffer`
+
+* params:
+  - `i32 (proxy_buffer_type_t) buffer_type`
+  - `i32 (offset_t) offset`
+  - `i32 (size_t) size`
+  - `i32 (iovec_array_t) buffer_data`
+  - `i32 (size_t) buffer_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Sets content of the buffer `buffer_type` to the values from memory slices
+pointed by the provided list of vectors (`buffer_data`, `buffer_size`),
+replacing `size` bytes in the existing buffer, starting at the `offset`. Host
+environment must adjust the size of the destination buffer (e.g. shrink the
+buffer if target `size` is larger than `buffer_size`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `buffer_type`.
+
+Returns `InvalidMemoryAccess` when `buffer_data` and/or `buffer_size` point to
+invalid memory address.
+
+#### `proxy_get_map_values`
+
+* params:
+  - `i32 (proxy_map_type_t) map_type`
+  - `i32 (iovec_array_t) keys_data`
+  - `i32 (size_t) keys_size`
+  - `i32 (iovec_array_t[]*) return_map_data`
+  - `i32 (size_t*) return_map_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Returns map containing key-values for keys from the provided list of vectors
```

Comment:
If this map is going to be used for HTTP headers, then do we want to standardize what names are there? Envoy has a particular set of system-defined headers like :path, :method, etc, and the rest are just lower-cased versions of the HTTP header. Should the spec discuss this so that people can write modules that can theoretically support multiple hosts?